remote partition_manifest validation for topic recovery

andijcr opened the pull request:
Introduced a new function to perform metadata validation for a remote topic to be used during the process of topic recovery.

For each partition, checks if the manifest exists in the cloud and optionally checks if the file can be decoded and self-consistent up to N most recent segment_meta,

For each partition, the S3 cost is 1 HeadObject request for the partition_manifest OR 1Get + 1HeadObject request for N segment_meta.

in general, we could limit the depth of the check by
- num segments
- offset
- total time

num segments correlates directly with the speed of the operation while being deterministic.
offset could map better with what the user wants, but it's more challenging to implement with the current reverse iteration implementation of segment_meta_cstore.
total time, it's a strong guarantee, but the final result depends on the load of the system.

This PR implements a max_num_segment limit, as it's easier to reason about. A follow-up can implement the other two modes if there is a request for this.

The check is performed in parallel for each partition with a cap.

The result can be 

```
passed   // checks are on
missing_manifest // allowed for topics that did not have time to upload a manifest

failure    // decoding failure or metadata inconsistencies; should stop 
download_issue // cloud storage configuration error or service error; should stop
```

The validation mode is driven by a new (nullopt) topic property and a cluster-level default + force flag.

A new topic_property allows users to define validation manually by editing the topic_manifest.json file.
otherwise, cluster-level defaults will be used.

```cpp
    enum_property<model::recovery_validation_mode>
      cloud_storage_recovery_topic_validation_mode;
    property<uint16_t> cloud_storage_recovery_topic_validation_depth;
    property<bool> cloud_storage_recovery_topic_force_ovveride_cfg;
```

The last one is an escape hatch to override `recovery_checks` with the cluster defaults
 
Fixes https://github.com/redpanda-data/core-internal/issues/1138
Fixes https://github.com/redpanda-data/core-internal/issues/1139

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [X] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

### Improvements

* before creating a recovery topic, perform metadata validation on the cloud data to ensure that each partition can be recovered successfully

Lazin commented on src/v/cloud_storage/anomalies_detector.h:
```diff
@@ -48,15 +48,35 @@ class anomalies_detector {
         result& operator+=(result&&);
     };
 
+    using segment_depth_t = named_type<int32_t, struct segment_depth_t>;
```
is it OK to have both type name and tag type name match?

andijcr replied:
I think so but here it's def a typo

Lazin commented on src/v/cloud_storage/remote.cc:
```diff
@@ -912,14 +912,64 @@ remote::download_object(cloud_storage::download_request download_request) {
+      bucket, cloud_storage_clients::object_key{segment_path()}, parent);
+}
+
+ss::future<std::pair<download_result, manifest_format>>
```
nit: can this be a struct?

andijcr replied:
done

Lazin commented on src/v/cloud_storage/remote.cc:
```diff
@@ -912,14 +912,64 @@ remote::download_object(cloud_storage::download_request download_request) {
+             generate_partition_manifest_path(
+               ntp, rev_id, manifest_format::serde),
+             parent)
+      .then([&bucket, &parent, ntp = std::move(ntp), rev_id, this](
```
nit: I think it will be more readable with co-await

Lazin commented on src/v/config/configuration.cc:
```diff
@@ -1868,6 +1868,34 @@ configuration::configuration()
+      "`check_manifest_own_metadata_only`",
+      {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
+      10)
+  , cloud_storage_recovery_topic_force_ovveride_cfg(
```
typo

Lazin commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,374 @@
+    auto [initial_rev_id, num_partitions]
+      = assignable_config.cfg.properties.remote_topic_properties.value();
+
+    if (num_partitions <= 0) {
```
do we really need this check?

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/46492#018e5ba2-f06b-4254-944c-8f621456b8ca:
```
"rptest.tests.e2e_topic_recovery_test.EndToEndTopicRecovery.test_restore_with_config_batches.num_messages=2.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_fast2.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_missing_partition.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.cluster_config_test.ClusterConfigTest.test_rpk_export_import"
"rptest.tests.control_character_flag_test.ControlCharacterPermittedAfterUpgrade.test_upgrade_from_pre_v23_2.initial_version=.23.1.1"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_vcluster_id.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_size_based_retention.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_empty_segments.cloud_storage_type=CloudStorageType.ABS"
```
<!--cd503cc4-190d-441d-9879-90e82d377115-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=16915) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=16915) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: andijcr<br/>:x: nobody<br/><hr/>**nobody** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=16915) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=16915) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=16915) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: andijcr<br/>:x: nobody<br/><hr/>**nobody** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=16915) it.</sub>

andijcr commented:
force push to fix conflicts and restart ci

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+    // this unpack is to trigger a compilation error if a new field is added
+    // to cloud_storage::anomalies. If a new anomalies is added, it should
+    // be also handled here
+    auto const& [missing_partition_manifest, missing_spillover_manifests, missing_segments, segment_anomalies, ignored_1, ignored_2, ignored_3, ignored_4]
```
nit: this is a very long line. Surprised it wasn't automatically wrapped 👀

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+  cloud_storage::remote& remote,
+  retry_chain_node& rtc,
+  recovery_checks checks) -> ss::future<validation_result> {
+    using enum validation_result;
```
This is cool in a smaller scope, but for larger functions like this I think it's valuable to keep the explicit enums stated. For instance, it isn't immediately clear if `failure` is referring to a member variable that I missed, or an enum

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+          ntp,
+          rev_id,
+          get_path());
+        co_return failure;
```
nit: maybe it's worth renaming to `anomaly_detected` or something? That way it's less ambiguous how this is different from `download_issue`

andijcr replied:
good point

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+    if (anomalies.status == cloud_storage::scrub_status::failed) {
+        // download issue. could be a configuration issue or a s3 issue
+        vlog(
+          clusterlog.error,
+          "{}/{} ({}) manifest metadata check download failure",
```
nit: rather than repeating {}/{} everywhere, consider using a retry_chain_logger? It's already instantiated up there

andrwng replied:
Also, I think it's worth considering putting all of these methods into their own `topic_recovery_validator` class, that can own common members (ntp, revision, loggers, paths, etc). May also make this more easily unit testable, and would have a great place (in the header) to detail expectations about the return values of these functions

andijcr replied:
done

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -1912,6 +1912,34 @@ configuration::configuration()
+      "`check_manifest_own_metadata_only`",
+      {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
+      10)
+  , cloud_storage_recovery_topic_force_override_cfg(
+      *this,
+      "cloud_storage_recovery_topic_force_override_cfg",
+      "If True, force validation to ignore the recovery_checks topic property, "
+      "and instead use cluster config values "
+      "`cloud_storage_recovery_topic_validation_mode` and "
+      "`cloud_storage_recovery_topic_validation_depth`",
+      {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
+      false)
```
I'm curious what the motivation is here? Presumably, an admin would have control over the incoming properties, no?

andijcr replied:
an escape hatch to disable validation altogether (cloud_storage_recovery_topic_validation_mode=no_check)

or to help testing to force a validation mode

andrwng commented on src/v/cloud_storage/anomalies_detector.cc:
```diff
@@ -32,10 +32,10 @@ anomalies_detector::anomalies_detector(
 
 ss::future<anomalies_detector::result> anomalies_detector::run(
   retry_chain_node& rtc_node,
-  archival::run_quota_t quota,
+  quota_limit quota_total,
   std::optional<model::offset> scrub_from) {
     _result = result{};
```
Also, I'm wondering about the choice to expose the number of cloud operations as the units with which to limit. On the one hand, it kind of makes sense as a tool to limit throughput (though if that's the goal, should it be expressed in some bytes?). On the other hand, segments aren't exposed to users. I wonder if expressing this as some number of offsets would make sense

andijcr replied:
the requisite expressed in the RFC is to express a limit for recovery checks as the number of segments, to control the overall duration of the operation. 

in general we could limit by
- num segments
- offset
- total time

num segments correlate directly with the speed of the operation while being deterministic.
offset could map better with what the user wants, but we would still need a total time limit i suppose.
total time it's a strong guarantee but the final result depends on the load of the system.

as a followup we could pass an offset limit and a total time limit too, and let the user decide.
.

I'll update the CL.

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -1912,6 +1912,34 @@ configuration::configuration()
+       .visibility = visibility::tunable},
+      model::recovery_validation_mode::check_manifest_existence,
+      {model::recovery_validation_mode::check_manifest_existence,
+       model::recovery_validation_mode::check_manifest_own_metadata_only,
```
nit: what does "own" refer to?

andijcr replied:
terrible name, changed to check_manifest_and_segment_metadata

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -1912,6 +1912,34 @@ configuration::configuration()
+  , cloud_storage_recovery_topic_validation_depth(
+      *this,
+      "cloud_storage_recovery_topic_validation_depth",
+      "Depth of metadata validation, when "
```
nit: "Depth" isn't a well-defined term in redpanda yet. Could you elaborate on what this means as it pertains to remote segments?

andijcr replied:
how many segments, from the newest to the oldest.
it's the third most difficult problem in cs: does a stack grow upward or downward? 😁

changed wording.

andrwng reviewed (commented):
The high level structure seems reasonable. Left a bunch of nits, and some questions about how this is exposed via configs.

Also I think the topic recovery validator and new bits in the anomaly detector could use some unit testing.

andijcr commented on src/v/cloud_storage/anomalies_detector.cc:
```diff
@@ -195,7 +193,28 @@ anomalies_detector::check_manifest(
+    // limit the total number (per partition) of visited segments in a run:
+    // first compute how many segments we can visit with the
+    // assigned_segment_arity(),
+    auto const visitable_segments_quota = std::min(
+      ptrdiff_t(manifest.size()), ptrdiff_t(get_visitable_segments()()));
+    // next position the start iterator to visit only the last
+    // #visitable_segments_quota. seg_iter.index() range: [0, manifest.size()]
+    auto seg_iter = std::next(
+      manifest.begin(), ptrdiff_t(manifest.size()) - visitable_segments_quota);
```
could we compromise with

```suggestion
    // next, visit only the last #visitable_tail_segments.
    // take begin iterator and skip the first `size() - visitable_tail_segments` elements.
    auto visitable_tail_segments = std::min(manifest.size(), get_visitable_segments()());
    auto start_index = manifest.size() - visitable_tail_segments;
    auto seg_iter = std::next(manifest.begin(), ptrdiff_t(start_index));
```

andijcr commented on src/v/cloud_storage/anomalies_detector.cc:
```diff
@@ -195,7 +193,28 @@ anomalies_detector::check_manifest(
-    auto seg_iter = manifest.begin();
+    // limit the total number (per partition) of visited segments in a run:
+    // first compute how many segments we can visit with the
+    // assigned_segment_arity(),
```
get_visitable_segments(). fixed

andijcr commented on src/v/cloud_storage/anomalies_detector.h:
```diff
@@ -49,14 +49,36 @@ class anomalies_detector {
 
+    using segment_depth_t = named_type<int32_t, struct segment_depth_tag>;
+    struct quota_limit {
+        archival::run_quota_t gets_quota = archival::run_quota_t::max();
```
done, changed the name for both.
the only snag is that "max_num_operations" is slightly misleading, the scrubber will at least process 1 to ensure forward progress, so max_num_operations=0 will still perform some calls.

andijcr commented on src/v/cluster/topic_recovery_validator.h:
```diff
@@ -0,0 +1,38 @@
+#pragma once
```
done

andijcr commented on src/v/cluster/topic_recovery_validator.h:
```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include "cloud_storage/remote.h"
+#include "cluster/types.h"
```
it's just used in cluster, it has less exposure here. should i move it?

andijcr commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -42,6 +42,7 @@
 #include "rpc/errc.h"
 #include "rpc/types.h"
 #include "ssx/future-util.h"
+#include "topic_recovery_validator.h"
```
done

andijcr commented on src/v/ssx/future-util.h:
```diff
@@ -295,11 +282,31 @@ inline bool is_shutdown_exception(const std::exception_ptr& e) {
+    } catch (const seastar::broken_promise&) {
+    } catch (const seastar::broken_condition_variable&) {
+    } catch (const seastar::nested_exception& e) {
+        if (!(is_shutdown_exception(e.outer)
+              && is_shutdown_exception(e.inner))) {
+            throw e;
+        }
```
removed. i observed a ci failure with this kind of shutdown exception but i have another pr just for this

dotnwat reviewed (commented):
minor comments/questions.

how do you feel about test coverage and what testing do you think we need to do?

agree with pretty much all of andrews comments as well.

andijcr commented on src/v/model/metadata.h:
```diff
@@ -618,6 +618,21 @@ struct broker_v0 {
 } // namespace internal
+
+enum class recovery_validation_mode : std::uint8_t {
+    check_manifest_existence
+    = 0, // ensure that either the manifest is in TS or that no manifest is
+         // present. download issues will fail the validation
+    check_manifest_own_metadata_only
+    = 1, // download the manifest and check the most recent segments up to
+         // max_segment_depth
+    no_check = 0xff, // do not perform any check, validation is considered
+                     // successful
```
fixed

andijcr commented on src/v/model/model.cc:
```diff
@@ -531,4 +531,33 @@ write_caching_mode_from_string(std::string_view s) {
+    case no_check:
+        return os << "no_check";
+    }
+    return os << "{unknown validation_mode}";
```
fixed

andijcr commented on src/v/model/metadata.h:
```diff
@@ -618,6 +618,21 @@ struct broker_v0 {
+                     // successful
+};
+
+auto operator<<(std::ostream&, recovery_validation_mode) -> std::ostream&;
+auto operator>>(std::istream&, recovery_validation_mode&) -> std::istream&;
```
done

andijcr commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+            // only res==success (manifest found) or res==notfound (manifest NOT
+            // found) make sense. warn for timedout and failed
+            if (download_res == cloud_storage::download_result::success) {
+                vlog(clusterlog.info, "{}/{} manifest found", ntp, rev_id);
```
my reasoning is that during recovery I want to see a trace of the recovery, 
so that we get

INFO ... Performing validation {} on topic {} with {} partitions
and for each partition either an INFO line or an ERROR line

dotnwat replied:
This makes sense. I think that the message should say something more than "manifest found", though. Put it in context--in reality there might be other stuff intermingled in the log, and "manifest found" isn't very self-describing.

andijcr commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+
+#include <seastar/coroutine/as_future.hh>
+
+// TODO design alternative: it's possible to use this subclass of
+// abort_requested_exception to support early exit in validation. `struct
+// terminate_early_exception final : public ss::abort_requested_exception {};`
+// do_validate_recovery_partition could terminate siblings operating on
+// different partisions like this:
+// `as.request_abort_ex(terminate_early_exception{})`
+// and while handling remote request, terminate_early_exceptions should be
+// flattened to a generic "eary_termination"  like so `co_await
+// remote.manifest_exists(bucket, path,
+// rtc).handle_exception_type([](terminate_early_exception) { return
+// early_termination; });` on the caller side, this local abort_source will
+// propagate root abort requests and local terminate_early requests
+// `ss::abort_source validation_as{}; auto master_as_subscription =
+// as.subscribe([&] { validation_as.request_abort(); });` ultimately this design
+// is not implemented here, since there seems no concrete use case to terminate
+// early at this level. either it's a general shutdown case, or the caller code
+// can get a better sense of what's going on by knowning the result of all
+// partitions
```
old self documentantion, removed

andijcr commented on src/v/kafka/server/handlers/topics/types.cc:
```diff
@@ -139,6 +139,38 @@ get_tristate_value(const config_map_t& config, std::string_view key) {
 }
 
+static std::optional<cluster::recovery_checks>
+get_recovery_checks(const config_map_t& config) {
```
src/v/kafka/server/tests/types_conversion_tests.cc::test_recovery_validation_modes

andijcr commented:
force push: rebase, fixing merge conflicts, addressed some of the comments.

TBD: 
- content and level of log messages in topic_recovery_validator
- unit test for topic_recovery validator

waiting ci to check regessions

andijcr commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,376 @@
+    // performed externally
+    using enum validation_result;
+
+    return remote.partition_manifest_exists(bucket, ntp, rev_id, rtc)
+      .then(
+        [ntp, rev_id](cloud_storage::remote::partition_manifest_existence res) {
```
simple leaf function, only one suspension point. no other reason other than that

andijcr commented:
fixed merge conflict, added unit test for anomalies detector, structured topic_recovery_validator with an internal class to factor common parts

TBD: extend TopicRecoveryTest to test more combinations of recovery checks

andijcr commented:
fixed merged conflict, added test_prevent_recovery to check in ducktape multiple combinations of recovery checks for a topic

andijcr commented:
> how do you feel about test coverage and what testing do you think we need to do?

should be done, the anomalies detector has a new unit test covering the limit by segment depth, 
while topic_recovery_validator is tested via ducktape.

the test run validation by using cluster level properties or topic level properties, it removed a segment and checks that recovery fails/not fails

andijcr commented on src/v/cluster/types.cc:
```diff
@@ -362,7 +362,8 @@ std::ostream& operator<<(std::ostream& o, const topic_properties& properties) {
       "write_caching: {}, "
       "flush_ms: {}, "
-      "flush_bytes: {}}}",
+      "flush_bytes: {}}},",
```
fixed

andijcr commented:
last failure was interesting, i had to manually cast the `mode` enum to its unsigned type, otherwise it would not be deserializable as uint. Other enums are fine. i didn't investigate the reason (too many jumps), but i suspect that ODR violation might be the problem and i'm hitting the wrong impl  
from the fix:
```
inline void rjson_serialize(
  json::Writer<json::StringBuffer>& w, const cluster::recovery_checks& rc) {
    w.StartObject();
    // TODO investigate the reason. seems like a manually casting to uint16_t of
    // rc.mode enum is needed, otherwise we get an assertion at decoding time,
    // when we try to read an unsigned but the json value is not tagged as
    // unsigned
    write_member(
      w,
      "mode",
      static_cast<std::underlying_type_t<decltype(rc.mode)>>(rc.mode));
    write_member(w, "max_segment_depth", rc.max_segment_depth);
    w.EndObject();
}
```

andijcr commented:
https://github.com/redpanda-data/redpanda/issues/17198 issue is known

andrwng commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -487,6 +489,32 @@ ss::future<topic_result> topics_frontend::do_create_topic(
               "valid "
               "topic manifest");
         }
+        auto validation_map = co_await maybe_validate_recovery_topic(
+          assignable_config, bucket, _cloud_storage_api.local(), _as.local());
+        if (std::ranges::any_of(
```
Should we consider the timeout here too?

andijcr replied:
you mean the timeout for the create_topics operations? 
yes we should (https://github.com/redpanda-data/core-internal/issues/1221) 

in the ducktape local environment, and from some testing also on a cluster with s3, the current defaults seems to be ok.

setting the cluster default validation to no_check would work as an escape hatch in case this operation becomes the bottleneck.

maybe i can return an optional<map<partition_id, validation_res>> to remove the runtime cost in case of nocheck

andrwng reviewed (commented):
In my last review I hadn't considered adding the option to the create request rather than the topic property, and the more I think about it the more natural it seems. WDYT?

dotnwat commented on src/v/cloud_storage/remote.cc:
```diff
@@ -988,6 +988,39 @@ ss::future<download_result> remote::segment_exists(
       existence_check_type::segment);
 }
 
+auto remote::partition_manifest_exists(
```
yeh, agree. i asked andrea to not worry about updating these in place, and update them with a new commit on the end or a separate pr.

dotnwat commented on src/v/cluster/types.h:
```diff
@@ -1582,6 +1598,11 @@ struct topic_properties
+    // introduced to drive validation at recovery time, when .recovery=true and
+    // .remote_topic_properties.has_value() the main use case is to have a
+    // manually editable knob in the topic_manifest json file
+    std::optional<recovery_checks> recovery_checks{std::nullopt};
```
@andijcr this is a good question. can you motivate the decision?

andijcr replied:
removed as discussed.

some excerpts from the discussion, as reference: 

"
Originally the idea was to be able to control checks by editing the topic_manifest.json . But with topic_manifest moving to serde that's not really an option anymore, so i can see how there is no reason for them to be in topic_properties
"

"
They were in topic_properties because the design was to give the option to use a different validation strategy per topic, instead of having only one per cluster. i see topic_manifest as a serialized topic_configuration, so it had the added bonus (once everything was in place) of being able to set offline the validation strategy for a topic by editing it.
The original design had the idea of leaving part of the validation to the node recreating the final partition, but that part was removed in the end
"

"
[validation] is done in the shard creating the create_topic command, but one possible extension was to just perform a manifest existence check and leave a deeper scrub to the shard managing the partition
"

dotnwat commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,330 @@
+        // .bin but the value is just a hint of the HEAD request that
+        // generated the abnormal result)
+        op_logger_.error(
+          "manifest {} download error: download_result: {}, validation NOT OK",
+          get_path(),
```
> nit: please let's avoid logging things in all caps without good reason (specifically the "NOT"). The fact that this is an error log should be signal enough that this is important

@andijcr please fix. we brought this up in an earlier review.

andijcr replied:
done

dotnwat commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,330 @@
+
+namespace cluster {
+
+class partition_validator {
+public:
+    // Each partition gets a separate retry_chain_node and a logger tied to it.
+    // From this a common retry_chain_logger is created
```
> Do you think it's worth unit testing this partition validator on its own? Defining it in the .cc file seems to discourage testing

good point. @andijcr we want unit testing to be the first line of defense for testing new code.

however, if we have some testing that we feel like will be exercising this, i'm ok if we do this as a follow up rather than continuing to iterate on this large PR. if we don't think any of this code is getting exercised, then we need to do that.

andijcr replied:
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.[test_prevent_topic_recovery|test_many_partitions] are written to test this code. i'll get a unit test going like for anomalies_detector. the dimension are [validation mode, cardinality 4], [partition damage, cardinality ~6], [download results, cardinatily ~3]

andijcr commented:
as discussed, removed recovery_checks from topic_properites and updated the rest of the code. 

Not checks are performed based on cluster properties.

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,307 @@
+    case check_manifest_and_segment_metadata:
+        return do_validate_manifest_metadata();
+    }
+
+    // Unhandled cases, compilation should prevent this. Ensure some
+    // level of forward compatibility
+    op_logger_.warn("{} not handled", checks_.mode);
+    return ss::make_ready_future<validation_result>(validation_result::passed);
```
nit: could just use `__builtin_unreachable`? Regarding forward compatibility, wouldn't we fail to serialize the cluster property and never begin validation?

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,307 @@
+    constexpr static auto is_fatal_anomaly =
+      [](cloud_storage::anomaly_meta const& am) {
+          using enum cloud_storage::anomaly_type;
+          switch (am.type) {
+          case missing_delta:
+          case non_monotonical_delta:
+          case end_delta_smaller:
+          case committed_smaller:
+              return true;
+          case offset_gap:
+          case offset_overlap:
+              return false;
```
At least from the metrics, there's an existing categorization of "high" and "low' severity, but it doesn't quite seem to match https://github.com/redpanda-data/redpanda/blob/24e7b780da6b0ac31174935def6e175d004652b2/src/v/cluster/partition_probe.cc#L368-L443. Is there a different categorization for what constitutes fatal?

andijcr replied:
changed to match

andrwng commented on src/v/cluster/topic_recovery_validator.cc:
```diff
@@ -0,0 +1,307 @@
+
+ss::future<validation_result>
+partition_validator::do_validate_manifest_existence() {
+    auto [download_res, _] = co_await remote_->partition_manifest_exists(
```
Shouldn't we use the format to log the correct path?

andijcr replied:
right

andrwng commented on tools/offline_log_viewer/controller.py:
```diff
@@ -106,9 +106,18 @@ def read_topic_properties_serde(rdr: Reader, version):
+            'flush_bytes':
+            rdr.read_optional(Reader.read_int64),
+            'recovery_checks':
+            rdr.read_optional(lambda rdr: rdr.read_envelope(
+                lambda rdr, _: {
+                    'mode': rdr.read_serde_enum(),
+                    'max_segment_depth': rdr.read_uint16()
+                }))
```
No longer needed?

andijcr replied:
right

andijcr commented:
updated on comments, the major change is match partition_probe to determine which segment_meta anomaly to consider fatal
